File: built-in plugins dsp/mb-dist-harmonics.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbDistHarmonicsPlugin: PluginDefinition = { id: 'mb-dist-harmonics', slug: 'mb-dist-harmonics', name: 'MB Harmonic Exciter', category: 'effect', type: 'distortion', version: '1.0.0', description: 'Harmonic enhancement', author: 'Max Booster', parameters: [{ id: 'odd', name: 'Odd', type: 'float', defaultValue: 0.3, minValue: 0, maxValue: 1, automatable: true }, { id: 'even', name: 'Even', type: 'float', defaultValue: 0.3, minValue: 0, maxValue: 1, automatable: true }, { id: 'mix', name: 'Mix', type: 'float', defaultValue: 0.5, minValue: 0, maxValue: 1, automatable: true }], defaultPreset: { odd: 0.3, even: 0.3, mix: 0.5 } };

export default MbDistHarmonicsPlugin;

File: built-in plugins dsp/mb-dist-harmonics.cpp
/**
 * MB Harmonic Exciter
 * Category : effect
 * Type     : distortion
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Harmonic enhancement
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_DIST_HARMONICS_H
#define MB_DIST_HARMONICS_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbDistHarmonics : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-dist-harmonics";
    static constexpr const char* PLUGIN_NAME    = "MB Harmonic Exciter";
    static constexpr const char* PLUGIN_TYPE    = "distortion";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float odd = 0.3f;  // range [0, 1]
    float even = 0.3f;  // range [0, 1]
    float mix = 0.5f;  // range [0, 1]
    };

    MbDistHarmonics() = default;
    ~MbDistHarmonics() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.odd = std::clamp(params.odd, 0f, 1f);
        params.even = std::clamp(params.even, 0f, 1f);
        params.mix = std::clamp(params.mix, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Harmonic Exciter
        return input;
    }
};

#endif // MB_DIST_HARMONICS_H
